### PR TITLE
Fix Version based BWC and set correct minCompatVersion

### PR DESCRIPTION
**Pull request by s1monw**
Approaching the release of 6.0 we need to sort out the usage of
`Version#minimumCompatibilityVersion` which was still set to 5.0.0.
Now this change moves it to the latest released version of 5.x (5.4 at this point)
to ensure we are compatible with the latest minor of the previous major. This change
also removes all the `_UNRELEASED` from the versions that where released and drops versions
that were never released and are not expected to be released (bugfixes in minors that are not
the latest in the previous major).

I will open separate PRs for 5.x

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 1]
```diff
@@ -438,11 +431,7 @@ static ReplicaItemExecutionMode replicaItemExecutionMode(final BulkItemRequest r
         } else {
             // NOTE: write requests originating from pre-6.0 nodes can send a no-op operation to
             // the replica; we ignore replication
-            // TODO: remove noOp result check from primary response, when pre-6.0 nodes are not supported
-            // we should return ReplicationItemExecutionMode.NORMAL instead
-            return primaryResponse.getResponse().getResult() != DocWriteResponse.Result.NOOP
-                    ? ReplicaItemExecutionMode.NORMAL // execution successful on primary
-                    : ReplicaItemExecutionMode.NOOP; // ignore replication
+            return ReplicaItemExecutionMode.NORMAL;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: @bleskes please take a look at this

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 2]
```diff
@@ -414,13 +414,6 @@ private static BulkItemResultHolder executeUpdateRequest(UpdateRequest updateReq
         FAILURE
     }
 
-    static {
-        assert Version.CURRENT.minimumCompatibilityVersion().after(Version.V_5_0_0) == false:
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: Do you want to remove the stuff in the TODO below like the note asks? I don't think we're likely remember to remove it unless you do it now or open an issue to do it as a followup to this PR.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: Or do it as a direct followup, I suppose.

File: core/src/main/java/org/elasticsearch/Version.java [hunk 3]
```diff
@@ -296,7 +280,7 @@ public Version minimumCompatibilityVersion() {
         final int bwcMinor;
         if (this.onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
             bwcMajor = major - 1;
-            bwcMinor = 0; // TODO we have to move this to the latest released minor of the last major but for now we just keep
+            bwcMinor = 4;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/Version.java]: Should we set this to 5 (for 5.5.0 now) and also add enforcement that it's increased any time a new minor version is bumped (e.g., to 6 if the version in 5.x is bumped)?

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 4]
```diff
@@ -1324,9 +1324,16 @@ public final void messageReceived(BytesReference reference, Channel channel, Str
                 }
                 streamIn = compressor.streamInput(streamIn);
             }
-            if (version.isCompatible(getCurrentVersion()) == false) {
+            Version compatibilityVersion = TransportStatus.isHandshake(status) ?
+                // for handshakes we are compatible with N-2 since otherwise we can't figure out our initial version
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: I think this'd be more readable if the comment was above the variable. I don't know enough about the handshake to verify the change off hand though.

File: core/src/main/java/org/elasticsearch/Version.java [hunk 5]
```diff
@@ -55,33 +55,25 @@
     public static final Version V_5_0_1 = new Version(V_5_0_1_ID, org.apache.lucene.util.Version.LUCENE_6_2_1);
     public static final int V_5_0_2_ID = 5000299;
     public static final Version V_5_0_2 = new Version(V_5_0_2_ID, org.apache.lucene.util.Version.LUCENE_6_2_1);
-    public static final int V_5_0_3_ID_UNRELEASED = 5000399;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/Version.java]: I think that we can enforce that these are marked as released in the same way that we enforce the list of upgrade versions in the packaging tests is bumped any time there is a release.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 6]
```diff
@@ -1324,9 +1324,16 @@ public final void messageReceived(BytesReference reference, Channel channel, Str
...
+                // since we are compatible with N-1 and N+1 so we always send our minCompatVersion as the initial version in the
+                // handshake. This looks odd but it's required to establish the connection correctly we check for real compatibility
+                // once the connection is established
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: ++

**Review by jasontedor — COMMENTED** [review]
I left two suggestions and question, the rest looks good.

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: I removed it and spoke to @bleskes he doens't know what it means and I have a test failure so I am just leaving it as it is.

File: core/src/main/java/org/elasticsearch/Version.java (hunk 5 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/Version.java]: ++ followup IMO

File: core/src/main/java/org/elasticsearch/Version.java (hunk 3 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/Version.java]: I am not sure we can / should do this. We guarantee that we are compatible with the latest minor of the previous major. That is `5.4` at this point. so I think this is correct? I think we also need to test against it though... it's a problem IMO but not a massive one.

File: core/src/main/java/org/elasticsearch/Version.java (hunk 5 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/Version.java]: I agree on follow-up.

File: core/src/main/java/org/elasticsearch/Version.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/Version.java]: I see, this is tricky.

**Review by jasontedor — APPROVED** [review]
LGTM.

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: I did some research and I think I get why this assertion was put in there. This assertion should check for V_6_0_0. That said, I'm not even sure we should have it and the accompanying TODO. I suggest changing the assertion to V_6_0_0 in this PR and I will work with @dakrone to clarify this in a follow up.

**Pull request merged**